src/parse/text/simple_extensions/argument.rs

```diff
@@ -0,0 +1,512 @@
+        }
+    }
+}
+
+impl<C: Context> Parse<C> for simple_extensions::ArgumentsItem {
+    type Parsed = ArgumentsItem;
+    type Error = ArgumentsItemError;
+
+    fn parse(self, ctx: &mut C) -> Result<Self::Parsed, Self::Error> {
+        match self {
+            simple_extensions::ArgumentsItem::Variant0 {
+                name,
+                description,
+                options,
+            } => EnumArgument::parse(name, description, options).map(Into::into),
+
+            simple_extensions::ArgumentsItem::Variant1 {
+                name,
+                description,
+                value,
+                constant,
+            } => ValueArgument::parse(ctx, name, description, value, constant).map(Into::into),
+
+            simple_extensions::ArgumentsItem::Variant2 {
+                name,
+                description,
+                type_,
+            } => TypeArgument::parse(ctx, name, description, type_).map(Into::into),
+        }
+    }
+}
+
+impl From<ArgumentsItem> for simple_extensions::ArgumentsItem {
+    fn from(value: ArgumentsItem) -> Self {
+        match value {
+            ArgumentsItem::EnumArgument(EnumArgument {
+                name,
+                description,
+                options,
+            }) => simple_extensions::ArgumentsItem::Variant0 {
+                name: Some(name),
+                description: Some(description),
+                options,
+            },
+
+            ArgumentsItem::ValueArgument(ValueArgument {
+                name,
+                description,
+                value,
+                constant,
+            }) => simple_extensions::ArgumentsItem::Variant1 {
+                name: Some(name),
+                description: Some(description),
+                value,
+                constant: Some(constant),
+            },
+
+            ArgumentsItem::TypeArgument(TypeArgument {
+                name,
+                description,
+                type_,
+            }) => simple_extensions::ArgumentsItem::Variant2 {
+                name: Some(name),
+                description: Some(description),
+                type_,
+            },
+        }
+    }
+}
+
+/// Parse errors for [simple_extensions::ArgumentsItem].
+#[derive(Debug, Error, PartialEq)]
+pub enum ArgumentsItemError {
+    /// Empty enumeration option.
+    #[error("empty enumeration option")]
+    EmptyEnumOption,
+
+    /// Empty optional field.
+    #[error("the optional field `{0}` is empty and should be removed")]
+    EmptyOptionalField(String),
+}
+
+/// Arguments that support a fixed set of declared values as constant arguments.
+#[derive(Clone, Debug, PartialEq)]
+pub struct EnumArgument {
+    /// A human-readable name for this argument to help clarify use.
+    name: String,
+
+    /// Additional description for this argument.
+    description: String,
+
+    /// List of valid string options for this argument.
+    options: Vec<String>,
+}
+
+impl EnumArgument {
+    fn parse(
```

Comment:
The approach you could use here is: implement `Parse` for the `ArgumentsItem` that forwards the parser to proceed parsing based on the variant. Those `Parse` implementations are parsers for the data structs of the variants (`EnumArgument`, `ValueArgument`, `TypeArgument`). I followed a similar approach for enums [for the `MappingType` enum (this was part of an unmerged PR)](https://github.com/substrait-io/substrait-rs/blob/aebd6cc53152cda3602decfbd096329e6f4e5850/src/parse/extensions/simple_extension_declaration/mod.rs).

Reply:
I'm still confused here.

The structure of `ArgumentsItem` contrasts with that of `MappingType`, as shown below. In `ArgumentsItem`, the enum fields are struct-like enum variants, whereas in `MappingType`, the enum fields consist of single-field tuples. So it is impossible to use the same approach that forwards the parser to proceed parsing based on the variant for `ArgumentsItem`. 

```
pub enum ArgumentsItem {
    Variant0 {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        description: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        name: Option<String>,
        options: Vec<String>,
    },
```
```
pub enum MappingType {
    #[prost(message, tag = "1")]
    ExtensionType(ExtensionType),
    #[prost(message, tag = "2")]
    ExtensionTypeVariation(ExtensionTypeVariation),
    #[prost(message, tag = "3")]
    ExtensionFunction(ExtensionFunction),
}
```

For `MappingType`, the code works:
```
match self {
    Self::ExtensionType(extension_type) => ctx.parse_record_value(extension_type)
```

However, for `ArgumentsItem`, the code throws compilation error:
```
error[E0164]: expected tuple struct or tuple variant, found struct variant `Self::Variant0`
  --> src/parse/text/simple_extensions/argument.rs:53:13
   |         match self {
53 |             Self::Variant0(enum_argument) => ctx.parse(enum_argument),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
```

So that, I have to destruct all the fields from the struct enum field and pass them to a `parse` function:
```
match self {
    simple_extensions::ArgumentsItem::Variant0 {
        name,
        description,
        options,
    } => EnumArgument::parse(name, description, options)
```

Correct me if I'm wrong.

Reply:
Oh right 🤦. Let me think a bit about this. Maybe we should try to change how these enums are generated by `typify`.

Reply:
It would be the best if we can change these enums.

Reply:
With the changes in https://github.com/substrait-io/substrait/pull/642, `typify` generates:
```rust
pub enum ArgumentsItem {
    EnumerationArg(EnumerationArg),
    ValueArg(ValueArg),
    TypeArg(TypeArg),
}

pub struct EnumerationArg {
    pub description: Option<String>,
    pub name: Option<String>,
    pub options: Vec<String>
}
```

Reply:
I will fix them once it merged.

Reply:
You can merge/rebase `main`.